pipeline/id_minter/src/main/scala/uk/ac/wellcome/platform/idminter/database/IdentifiersDao.scala

```diff
@@ -25,68 +25,73 @@ object IdentifiersDao {
 class IdentifiersDao(identifiers: IdentifiersTable) extends Logging {
   import IdentifiersDao._
 
-  def withConnection[R](doWork: DBSession => R): R =
-    DB localTx doWork
-
   def lookupIds(sourceIdentifiers: Seq[SourceIdentifier])(
-    implicit session: DBSession): Try[LookupResult] =
+    implicit session: DBSession = readOnlySession): Try[LookupResult] =
     Try {
       debug(s"Matching ($sourceIdentifiers)")
+      val distinctIdentifiers = sourceIdentifiers.distinct
       val identifierRowsMap: Map[(String, String, String), SourceIdentifier] =
-        sourceIdentifiers.map { id =>
+        distinctIdentifiers.map { id =>
           (id.ontologyType, id.identifierType.id, id.value) -> id
         }.toMap
 
-      blocking {
-
-        val i = identifiers.i
-        val matchRow = rowMatcherFromSyntaxProvider(i) _
-
-        // The query is manually constructed like this because using WHERE IN
-        // statements with multiple items causes a full-index scan on MySQL 5.6.
-        // See: https://stackoverflow.com/a/53180813
-        // Therefore, we perform the rewrites here on the client.
-        //
-        // Because of issues in `scalikejdbc`'s typing, it's necessary to
-        // specify the `Nothing` type parameters as done here in order for
-        // the `map` statement to work properly.
-        val query = withSQL[Nothing] {
-          select
-            .from[Nothing](identifiers as i)
-            .where
-            .map { query =>
-              sourceIdentifiers.tail.foldLeft(
-                query.withRoundBracket(matchRow(sourceIdentifiers.head))
-              ) { (q, sourceIdentifier) =>
-                q.or.withRoundBracket(matchRow(sourceIdentifier))
+      val foundIdentifiers =
+        withTimeWarning(thresholdMillis = 10000L, distinctIdentifiers) {
+          batchSourceIdentifiers(distinctIdentifiers)
+            .flatMap { identifierBatch =>
+              blocking {
+                val i = identifiers.i
+                val matchRow = rowMatcherFromSyntaxProvider(i) _
+
+                // The query is manually constructed like this because using WHERE IN
+                // statements with multiple items causes a full-index scan on MySQL 5.6.
+                // See: https://stackoverflow.com/a/53180813
+                // Therefore, we perform the rewrites here on the client.
+                //
+                // Because of issues in `scalikejdbc`'s typing, it's necessary to
+                // specify the `Nothing` type parameters as done here in order for
+                // the `map` statement to work properly.
+                val query = withSQL[Nothing] {
```

Comment:
Eww...Not saying you're doing anything wrong here but this is where a DSL is really not helping methinks. 

`<heresy>`Also possibly scala is not helping. Maybe the idminter would be simpler in python`</heresy>`

Reply:
fwiw this is old code that got indented differently... but yes, absolutely, and I remember when I originally wrote it it took _ages_ to work that out (and there were other issues which made it really hard to do with the library's `sql` and `sqls` string template things).

I agree with the heresy - in fact I'd go further and say it might be better/simpler in the language that's the J in JSON! 😱

Reply:
😱